inspirehep/modules/tools/authortool.py

```diff
@@ -0,0 +1,311 @@
+            elif aff_type == 'digit':
+                if not char.isdigit():
+                    return None
+        else:
+            if char.isalpha():
+                aff_type = 'alpha'
+            elif char.isdigit():
+                aff_type = 'digit'
+            else:
+                aff_type = 'symbol'
+                break
+    return aff_type
+
+def determine_aff_type(text):
+    """
+    Guess format for affiliations.
+    Return corresponding search pattern.
+    """
+
+    single_char = []
+    first_char = []
+    for line in text.split('\n'):
+        line = line.strip(' .')
+        if len(line) == 1:
+            single_char.append(line)
+        elif line:
+            first_char.append(line[0])
+
+    if single_char:
+        aff_type = determine_aff_type_character(single_char)
+        if aff_type:
+            aff_pattern = line_pattern_single[aff_type]
+        else:
+            raise ValueError('Cannot identify type of affiliation', single_char)
+    else:
+        aff_type = determine_aff_type_character(first_char)
+        if aff_type:
+            aff_pattern = line_pattern_line[aff_type]
+        else:
+            raise ValueError('Cannot identify type of affiliations', first_char)
+
+    return aff_pattern
+
+def parse_affiliations(text):
+    """
+    Determine how affiliations are formatted.
+    Return hash of id:affiliation
+
+    Allowed formats:
+    don't mix letters and numbers, lower-case letters only
+
+    1
+    CERN, Switzerland
+    2
+    DESY,
+    Germany
+
+
+    1 CERN, Switzerland
+    2DESY, Germany
+
+    a  CERN, Switzerland
+    bb DESY, Germany
+
+    *
+    CERN, Switzerland
+    #
+    DESY, Germany
+    """
+
+    affiliations = {}
+    aff_pattern = determine_aff_type(text)
+
+    aff_id = None
+    this_aff = []
+    for line in text.split('\n'):
+        line = line.strip()
+        get_affiliation = aff_pattern.search(line)
+        if get_affiliation:
+            if len(get_affiliation.groups()) == 2:
+                affiliations[get_affiliation.group(1)] = \
+                    get_affiliation.group(2).strip()
+            else:
+                if aff_id and this_aff:
+                    affiliations[aff_id] = ' '.join(this_aff).strip()
+                    aff_id = None
+                    this_aff = []
+                aff_id = get_affiliation.group(1)
+        elif aff_id:
+            this_aff.append(line)
+        elif line:
+            raise ValueError('Something is wrong with the affiliation list')
+    if aff_id and this_aff:
+        affiliations[aff_id] = ' '.join(this_aff)
+
+    return affiliations
+
+def create_authors(text):
+    """
+    Split text in (useful) blocks, sepatated by empty lines.
+    1 block: no affiliations
+    2 blocks: authors and affiliations
+    more blocks: authors grouped by affiliation (not implemented yet)
+
+    Return list of (author_fullname, [author_affiliations])
+    """
+
+    text = unicode(text, 'utf-8')
```

Comment:
This is making the tests fail. What are you trying to do?

Reply:
Test-case:
† 
CERN
*
DESY

I read the lines and get (print len(line), line)
3 †
4 CERN
1 *
4 DESY


after calling unicode I get
1 †
4 CERN
1 *
4 DESY

and I need  † to match regex '.'

Reply:
I can move unicode(text, 'utf-8') to the main code where I read the file, hoping it works with input from the web-form.

Reply:
This is happening because at the beginning you are dealing with a series of bytes containing UTF-8 encoded unicode. `len` is counting the number of bytes, but UTF-8 is a variable-length encoding, meaning that one character may be encoded either in one byte (ASCII) or several (the `†`). On the other hand calling `len` on unicode strings counts the number of characters.

What you need to do is call `decode` at the beginning, if `text` contains bytes and not unicode strings. So you should add `import six` (for python 3 compatibility) at the top, and add at the beginning of `parse_authors`:
```
if isinstance(text, six.binary_type):
    text = text.decode('utf-8')
```

Reply:
OK. I'll ask Florian to install six.